go/vt/vtgate/vstream_manager.go

```diff
@@ -798,20 +799,34 @@ func (vs *vstream) streamFromTablet(ctx context.Context, sgtid *binlogdatapb.Sha
 // An error should be retried if it is expected to be transient.
 // A tablet should be ignored upon retry if it's likely another tablet will not
 // produce the same error.
-func (vs *vstream) shouldRetry(err error) (bool, bool) {
+func (vs *vstream) shouldRetry(err error) (retry bool, ignoreTablet bool) {
 	errCode := vterrors.Code(err)
-
+	// In this context, where we will run the tablet picker again on retry, these
+	// codes indicate that it's worth a retry as the error is likely a transient
+	// one with a tablet or within the shard.
 	if errCode == vtrpcpb.Code_FAILED_PRECONDITION || errCode == vtrpcpb.Code_UNAVAILABLE {
 		return true, false
 	}
-
-	// If there is a GTIDSet Mismatch on the tablet, omit it from the candidate
-	// list in the TabletPicker on retry.
-	if errCode == vtrpcpb.Code_INVALID_ARGUMENT && strings.Contains(err.Error(), "GTIDSet Mismatch") {
-		return true, true
+	// This typically indicates that the user provided invalid arguments for the
+	// VStream so we should not retry.
+	if errCode == vtrpcpb.Code_INVALID_ARGUMENT {
+		// But if there is a GTIDSet Mismatch on the tablet, omit that tablet from
+		// the candidate list in the TabletPicker and retry. The argument was invalid
+		// *for that specific *tablet* but it's not generally invalid.
+		if strings.Contains(err.Error(), "GTIDSet Mismatch") {
+			return true, true
+		}
+		return false, false
+	}
+	// Internal errors such as not having all journaling partipants require a new
+	// VStream.
+	if errCode == vtrpcpb.Code_INTERNAL {
+		return false, false
 	}
 
-	return false, false
+	// For anything else, if this is a recoverable/ephemeral error -- such as a
+	// MAX_EXECUTION_TIME SQL error during the copy phase -- then retry.
+	return !vreplication.IsUnrecoverableError(err), false
```

Comment:
Should we move this to its own package or maybe to like `vterrors` so we don't pull in all of the `vreplication` package here?

Reply:
I don't mind doing that, but from when I've looked at the binaries built in cases like this the binary only includes the necessary symbols. 

Maybe you're thinking more about the package level management than the actual binaries though?

Reply:
> Maybe you're thinking more about the package level management than the actual binaries though?

More from a design perspective, but also even if it's not in the final binary, it stil compiles a lot more etc. And I think these are cases where having its own or putting it in a smaller more focused package is useful to do. It also makes it easier to reuse by avoiding things like import loops in the future which would force the same refactor then anyway too.

Reply:
No disagreement there. It's not quite general purpose, but I can move it to vterrors or somewhere else and include VReplication in the name.

Reply:
I was going to create a subdir and vreplerror package but after thinking about it more... we don't need to re-use the same function in the vstream manager as our context is a little different (e.g. it's read-only SELECTs in that context so a much smaller set of relevant errors). So I switched over to using a similar, but more generic, function that I had created for this purpose some time back: https://github.com/vitessio/vitess/pull/17783/commits/7c5c747242b1c701f8668a43d415a5628df680cd